Make `ContextHandler` catch and ignore exceptions thrown by `Thread.setContextClassLoader()`

lorban opened the pull request:
JDK25's `InnocuousThread` throws `SecurityException` when `Thread.setContextClassLoader()` is called with a non-null value.

If we want to support -for instance- an AIO thread calling into a context's request and response, we have to catch those exceptions.

See https://github.com/spring-projects/spring-framework/issues/35531 as well as https://github.com/jetty/jetty.project/pull/13669 for the original report.

lorban committed:
Ignore SecurityException thrown by setContextClassLoader() so JDK25's InnocuousThread can call into a context's request and response.

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -733,11 +733,23 @@ protected ClassLoader enterScope(Request contextRequest)
+        {
+            Thread.currentThread().setContextClassLoader(classLoader);
+        }
+        catch (SecurityException e)
```
Do the null check in the new method only.
That way you catch it for both enter & exit

```suggestion
    private void setContextClassLoader(ClassLoader classLoader)
    {
        if (classLoader == null)
            return;
        try
        {
            Thread.currentThread().setContextClassLoader(classLoader);
        }
        catch (SecurityException e)
```

sbordet replied:
I don't think we can skip the call if `classLoader==null`.

joakime replied:
Grrr...

The call `Thread.setContextClassaLoader(null)` has meaning, it's not just a no-op.

* https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Thread.html#setContextClassLoader(java.lang.ClassLoader)

> **Parameters**:
> `cl` - the context ClassLoader for this Thread, or null indicating the system class loader (or, failing that, the bootstrap class loader)

gregw replied:
I think we can skip the call.   If we say that a `ContextHandler` with a null ClassLoader has no requirement for its TCCL, then it is fine to skip the call.   If a context specifically needs the system or bootstrap class loader, then they can be explicitly set on the `ContextHandler`.

This is a slight change of meaning/behaviour - we make a distinction between a context that doesn't care about the classloader vs one that wants a specific one.

lorban replied:
This is a potentially breaking change, but I don't think anyone would be impacted by this new behavior which I think is saner.

Let me do this.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -733,11 +733,23 @@ protected ClassLoader enterScope(Request contextRequest)
+        {
+            Thread.currentThread().setContextClassLoader(classLoader);
+        }
+        catch (SecurityException e)
```
Just `catch(Throwable)`, so we won't have to come back in the future.

gregw replied:
No - this could be on the hot path.  We will run slow if we create exceptions every time

lorban replied:
If we throw here, we have a more serious problem than performance as (at least in the context of the newly added `testAIOAndClassLoader` test) the `Callback` passed to `Response.write()` is never completed.

gregw replied:
I'd rather break than run slow.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -760,7 +772,7 @@ protected void exitScope(Request request, Context lastContext, ClassLoader lastL
         notifyExitScope(request);
         __context.set(lastContext);
-        Thread.currentThread().setContextClassLoader(lastLoader);
+        setContextClassLoader(lastLoader);
```
We do not want to pay the cost if the the TCCL was not set, so this should be:


```suggestion
        if (_classLoader != null)
            setContextClassLoader(lastLoader);
```

gregw replied:
Either this or what @janbartel proposed

lorban replied:
Done.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ContextHandlerTest.java:
```diff
@@ -170,6 +175,80 @@ public void testMiss() throws Exception
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testAIOAndClassLoader(boolean withClassLoader) throws Exception
```
Can you move this test to `jetty-test-client-transports`?

I don't like mocking critical components that may themselves set the TCCL, so I'd prefer a full test with a real server and a real client.

lorban replied:
I prefer to keep this test as it is closer to the tested code, but I'll add a similar one to `jetty-test-client-transports` to make sure the real transports also behave properly.

sbordet replied:
There is no need to have duplicates.

janbartel commented:
How about:

```java
  protected void exitScope(Request request, Context lastContext, ClassLoader lastLoader)
    {
        notifyExitScope(request);
        __context.set(lastContext);
        if (lastLoader != Thread.currentThread().getContextClassLoader())
            Thread.currentThread().setContextClassLoader(lastLoader);
    }
```

In `enterScope()` we're not calling `setClassLoader()` if our context has a null classloader, so the `InnocuousThread`'s classloader will not have been changed when it calls `exitScope()` so we can skip the `setContextClassLoader()`.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -733,11 +733,23 @@ protected ClassLoader enterScope(Request contextRequest)
         return lastLoader;
     }
 
+    private void setContextClassLoader(ClassLoader classLoader)
```
```suggestion
    private void setThreadContextClassLoader(ClassLoader classLoader)
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -733,11 +733,23 @@ protected ClassLoader enterScope(Request contextRequest)
+        }
+        catch (SecurityException e)
+        {
+            // ignore
```
I do not think we can ignore this.  At best it will be wrong, at worst we will carry on with somebody else's classloader.

We should try to avoid the exception if we can, but if we hit it, then it is a hard fail

lorban commented:
@janbartel 's proposal is a more elegant version of my original proposal, but it only solves the problem when a null classloader is configured. When a non-null one is configured, `enterScope` throws.

lorban committed:
- Javadoc the behavior of what happens with the thread context's classloader, including when it is null
 - Never set the thread's context classloader when null is configured
 - Gracefully handle failures to set the thread's context classloader
 - Add an optimization for JDK25+ InnocuousThread

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
I've attempted to take all comments into account and come up with a safer & documented contract w.r.t thread context classloader, leaving it untouched if null is configured and playing it safe when we try to set it.

I've used @gregw's innocuous thread detection code but only used it as an optimization: if the detection fails for any reason, the only drawback is a slight perf hit as an exception is then thrown but safely caught.

lorban committed:
fix InnocuousThreadGroup detection

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add extra test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/jdk25-InnocuousThread-aio-ContextClassLoader

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadUtils.java:
```diff
@@ -0,0 +1,39 @@
+            return thread.getThreadGroup() == innocuousThreadGroup;
+
+        ThreadGroup threadGroup = thread.getThreadGroup();
+        if (threadGroup != null && "InnocuousThreadGroup".equals(threadGroup.getName()))
```
I am -1 on this check against private JDK classes that may be renamed, or changed.

The cases where we fail to set the TCCL are very restricted, and I would rather try/catch in those restricted cases than having to maintain this code over JDK versions.

`InnocuousThread` was introduced in JDK 9 12 years ago (https://github.com/openjdk/jdk/commit/115e92c184f1e75215cadc4b5a1100578d72b055), so the initial check about Java 25 is already possibly wrong.

lorban replied:
I personally am -0 on this. I don't mind adding implementation-specific checks as long as they are totally safe, which is the case here. But I do hear the argument that this is an extra annoyance to maintain just for a special, potentially rare use-case.

sbordet replied:
> I personally am -0 on this. I don't mind adding implementation-specific checks as long as they are totally safe, which is the case here. 

How are they safe? 
A rename of the JDK thread group would pass under the radar, and we will be back to this issue not being resolved.
And then we would have to sniff the JDK version, or have a list of names to compare against, and hope that when a name matches, it won't be for a thread for which it would be possible to set the TCCL.

For example, what happens with virtual threads?
My suspect is that JDK's AIO has not been updated to virtual threads internally, because it probably relies on user code to pass in a virtual thread `Executor` via `AsynchronousFileChannel.open(... Executor executor, ...)` (https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/nio/channels/AsynchronousFileChannel.html#open(java.nio.file.Path,java.util.Set,java.util.concurrent.ExecutorService,java.nio.file.attribute.FileAttribute...)), at which point all the bets on the thread group name are off.

The `try/catch` will work in all cases.

gregw replied:
I too an -0 on name checks.

But I'm -1 on ignoring any thrown exception on the hot path.   IF we are going to work when called by an Innocuous Thread, then we should work efficiently.     I'd much rather fail with a reason than work slowly.

If we need this method (and I currently don't think we do), we could perhaps avoid the name comparison if we could reliably create a single Innocuous Thread statically and take a copy of its ThreadGroup.

We should try to do that as we could then put some asserts into enter like:

```java
   assert _classloader == null || !ThreadUtils.isInnocuous(Thread.currentThread());
```

I  know people don't always run with asserts on, but it would give us an easier way to declare in the code where innocuous threads are not expected to work

gregw replied:
Ah! looking at this more, I do not think we can implement isInnocuous, as the JVM treats innocuous threads more of a concept rather than a specific type.   There are at least two separate implementations:
 + jdk.internal.misc.InnocuousThread
 + java.util.concurrent.ForkJoinWorkerThread.InnocuousForkJoinWorkerThread

Thus there is nothing to stop future implementations of the JVM to add more Innocuous Threads implementations and/or change the names.   There is no single ThreadGroup we can check and no interface.

I think they should never have let these threads into the wild!  I get it that you might want a very restricted thread to be used in a callback in some circumstances, but what they are going to cause is that every callback that calls a 3rd party library now will have to do an extra dispatch.    That would be fine if innocuous threads were really cheap, but they are not.

I think we just need to recommend that a thread pool always be used with APIs that might otherwise use innocuous threads.

gregw replied:
@lorban @sbordet see above

sbordet replied:
@gregw I agree about the fact that we cannot implement `isInnocuous()`.

Having said that, we have 2 choices:

1. Try to set the TCCL, catch, and hope that without the TCCL whatever code needed it works.
2. Try to fail hard.

Solution 2 seems to be quite complicated due to the fact that enterScope/exitScope are implemented for generic lambdas such as predicates, runnables, consumers, etc.
If we cannot enterScope, throwing would likely result in some unexpected behavior and the lambda would not be invoked.
For example, the predicate case is called on an idle timeout event, and throwing because we cannot enterScope would basically result in the idle timeout being ignored.

I'm for solution 2. but it requires a large review.

gregw replied:
I'm for option 3. Don't set a context classloader if the context doesn't have a classloader

gregw replied:
Or option 4.  Have a special NO_CLASS_LOADER constant ClassLoader that if set on a context stops it setting/resetting the TCCL - otherwise we keep our behaviour the same.
Those wanting to call us with InnocuousThreads will have to set NO_CLASS_LOADER - a PITA but perhaps better than a behaviour change.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -728,14 +734,39 @@ protected void initializeDefaultsComplete()
+        if (ThreadUtils.isInnocuous(Thread.currentThread()))
+        {
+            lastLoader = null;
+        }
```
I'd remove this.

lorban replied:
If the `isInnocuous()` thread util goes, this obviously goes too.

gregw replied:
I don't think this is correct here.
At the very least, we should avoid throwing and exception unless the context classloader is null.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -756,11 +787,20 @@ protected void notifyEnterScope(Request request)
+        notifyExitScope(contextRequest);
         __context.set(lastContext);
-        Thread.currentThread().setContextClassLoader(lastLoader);
+        if (lastLoader != null && lastLoader != Thread.currentThread().getContextClassLoader())
```
```suggestion
        if (_classLoader != null)
```

lorban commented:
@gregw @sbordet we have multiple problems here:

1. Attempting to call `ContextHandler`'s `Response.write()` from an `InnocuousThread` fails by throwing from `exitScope()` even when no classloader is configured. This can be fixed by not calling `setContextClassLoader()` when the `ContextHandler` has no configured classloader.
2. Attempting to call `ContextHandler`'s `Response.write()` from an `InnocuousThread` when the `ContextHandler` does have a configured classloader is badly buggy as the callback is never completed.
3. Do we want to allow writes to a context-handler-with-a-classloader's response from an `InnocuousThread`? We may decide that we don't as this is a rather uncommon use-case that could be worked around by the user (e.g.: dispatch the write to another thread, or do not configure a classloader on the `ContextHandler`) as long as we behave sanely. Or we can catch the exception in certain cases/all cases to allow such writes, ignoring the fact that we could not set the classloader.

(1) is a no-brainer: cheap and easy to fix (for instance, using Jan's proposal), totally solves the problem when no extra classloader is involved. Let's not discuss this as this will be done.
(2) is quite serious and more complicated to fix as there are invokers involved. Depending on what we do for (3) this has to be fixed or not.
(3) is an opinionated decision we have to take: do we allow those very special writes or not? I'm slightly leaning for NOT supporting them which means fixing (2) by making sure the callback gets failed, but I could be convinced otherwise.

lorban committed:
remove innocuous thread detection + also catch exceptions on TCCL restoration

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
This issue just became more burning as Spring will start recommending JDK25 with Boot 4 / Framework 7 to be released in November: https://github.com/jetty/jetty.project/pull/13669#issuecomment-3401545480

If we can't come up with a complete fix shortly, I'll split this issue in two:
 - fixing the problem when the ContextHandler's classloader is null (straightforward; solves the originally reported case)
 - fixing the problem when the ContextHandler's classloader is non-null

The former should get most of the heat away and will leave us more time to come up with a satisfying solution for the latter.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ContextHandlerTest.java:
```diff
@@ -170,6 +175,150 @@ public void testMiss() throws Exception
+            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            {
+                AtomicReference<Throwable> failureRef = new AtomicReference<>();
+                request.addFailureListener(failureRef::set);
```
Is this necessary?

lorban replied:
Not anymore, removed.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ContextHandlerTest.java:
```diff
@@ -170,6 +175,150 @@ public void testMiss() throws Exception
+            public boolean handle(Request request, Response response, Callback callback) throws Exception
+            {
+                AtomicReference<Throwable> failureRef = new AtomicReference<>();
+                request.addFailureListener(failureRef::set);
```
Is this necessary?

lorban replied:
Not anymore, removed.

sbordet commented:
After discussion, I think the solution could be as follows:

1. we try to set the TCCL; if we fail, we catch and ignore. Then we have 2 cases:
   1. The subsequent operation does not need the TCCL, and succeeds.
   2. The subsequent operation needs the TCCL and fails; the stack trace will show it comes from an innocuous thread and we can document/tell how to fix it.

The solution above may always throw/catch/ignore.

To work that around, it would be possible to have a `ThreadLocal<Boolean>` `ContextHandler.TCCL_SET` where we store whether we succeeded/failed to set the TCCL.
This value will never be cleared for that thread, but it's a little memory overhead with respect to the cost of throwing an exception every time.
We can skip setting it if it's a virtual thread? Unless the virtual thread is pooled 😂

lorban committed:
rename TL

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not rely on member variable in exitScope() as it can be nulled before exiting the scope

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use LocalConnector instead of mocks

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -756,11 +782,31 @@ protected void notifyEnterScope(Request request)
+        CURRENT_CONTEXT.set(lastContext);
+
+        notifyExitScope(contextRequest);
+        if (lastLoader != null && Thread.currentThread().getContextClassLoader() != lastLoader)
```
FYI: this cannot check `_classLoader` as that variable can be (and is, see: `WebAppContextTest`) set to null before `exitScope()` is called.

gregw replied:
We should always reset to `lastLoader` if `_classLoader!=null`.  Even if `lastLoader` is null.

sbordet replied:
@gregw we can't as `_classLoader` can be nulled out in between `enterScope()` and `exitScope()` -- there is a test that does that.
We can only rely on `lastLoader`.

gregw replied:
I do not believe it is valid to change the _classloader whilst a thread is running in the context.
However, I'll see if I can handle that bizarre case in my branch....

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -728,14 +733,35 @@ protected void initializeDefaultsComplete()
+                Thread.currentThread().setContextClassLoader(_classLoader);
+            }
+            catch (Throwable x)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("error setting a context classloader on thread {}", Thread.currentThread(), x);
+                lastLoader = null;
+            }
```
We should fail here rather than ignore

gregw commented:
The conversation is a bit split between multiple threads, so I'll repeat my thoughts here:

We need a mode where the context does not try to set/reset the TCCL at all.   This can be triggered by _classLoader == null (a behaviour change) or using a NO_CLASS_LOADER constant (a small change needed in spring code).
If a context has a normal classloader, then we have to solve that a different way - or just fail hard

lorban committed:
update according to latest discussion

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
update javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

Lublanski commented:
@lorban @gregw @sbordet 
I'm glad the reported issue has been resolved. I'm just a little disappointed because I was the one who filed the initial report and expressed interest in trying to contribute to an open-source project like Jetty.

Anyway, I'll look for another, simpler issue to try to make my first contribution. Congratulations on the fix, everyone.

sbordet commented:
@Lublanski sorry if this issue did not go your way... 
Please take a look at the issues with the ["Help Wanted" label](https://github.com/jetty/jetty.project/issues?q=state%3Aopen%20label%3A%22Help%20Wanted%22) if you would like to contribute.
Thanks!

joakime commented:
> I'm glad the reported issue has been resolved. I'm just a little disappointed because I was the one who filed the initial report and expressed interest in trying to contribute to an open-source project like Jetty.

We reviewed your proposed PR and found that it was creative, solved the problem, and would have likely worked.
But it also had some negatives, ones that were discussed, and the path to an eventual solution was quite varied (at least 3 different approaches), the final solution is one that we are happy to continue to support.

The biggest impact you have had on this project is issue #13666 that you filed!
Those details surprised us!  But it got a lot of us invested in finding an appropriate solution.
So thank you for that issue.

lorban commented:
@Lublanski The simple fact that you took the time to file this issue and to create a candidate fix is already extremely appreciated. Thanks for that, bug reports of good quality are invaluable to every open-source project.

Lublanski commented:
@sbordet @lorban @joakime 
I'm glad I was able to contribute in some way. Thank you guys.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -728,12 +733,31 @@ protected void initializeDefaultsComplete()
+            }
+            catch (Throwable x)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("error setting a context classloader on thread {}", Thread.currentThread(), x);
```
I know this has been merged, 
but I think we should make this a warn the first time it happens.
